NeoML/src/TraditionalML/Svm.cpp

```diff
@@ -49,71 +145,32 @@ CPtr<IModel> CSvm::Train( const IProblem& problem )
 		return nullptr;
 	}
 
-	CSvmKernel kernel( params.KernelType, params.Degree, params.Gamma, params.Coeff0 );
+	const CSvmKernel kernel( params.KernelType, params.Degree, params.Gamma, params.Coeff0 );
 
-	CSMOptimizer optimizer( kernel, problem, params.MaxIterations, params.ErrorWeight, params.Tolerance,
-		params.DoShrinking );
+	CSMOptimizer optimizer( kernel, problem, params.MaxIterations,
+		params.ErrorWeight, params.Tolerance, params.DoShrinking );
 	if( log != nullptr ) {
 		optimizer.SetLog( log );
 	}
 
-	CArray<double> alpha;
-	float freeTerm;
+	CArray<double> alpha{};
+	float freeTerm{};
 	optimizer.Optimize( alpha, freeTerm );
 
 	if( kernel.KernelType() == CSvmKernel::KT_Linear ) {
-		const int vectorCount = problem.GetVectorCount();
-		const int curThreadCount = IsOmpRelevant( vectorCount ) ? params.ThreadCount : 1;
-		const CFloatMatrixDesc matrix = problem.GetMatrix();
-		NeoAssert( matrix.Height == problem.GetVectorCount() );
-		NeoAssert( matrix.Width == problem.GetFeatureCount() );
-
-		CArray<CFloatVector> planeReduction;
-		planeReduction.Add( CFloatVector( problem.GetFeatureCount() + 1, 0.f ), curThreadCount );
-
-		NEOML_OMP_NUM_THREADS( curThreadCount )
-		{
-			const int threadNumber = OmpGetThreadNum();
-			CFloatVector& plane = planeReduction[threadNumber];
-
-			int index = 0;
-			int count = 0;
-			if( OmpGetTaskIndexAndCount( vectorCount, index, count ) ) {
-				for( int i = 0; i < count; i++ ) {
-					float alphaValue = static_cast<float>( alpha[index] * problem.GetBinaryClass(index) );
-					CFloatVectorDesc desc;
-					matrix.GetRow( index, desc );
-					plane.MultiplyAndAdd( desc, alphaValue );
-					index++;
-				}
-			}
-		}
+		CFindPlanesThreadTask planeTask( params.ThreadCount, problem, alpha );
+		NEOML_NUM_THREADS( *threadPool, &planeTask, []( int threadIndex, void* ptr ) {
+			( ( IThreadTask* )ptr )->CallRun( threadIndex );
+		} );
+		CFloatVector plane = planeTask.Reduction( freeTerm );
 
-		CFloatVector plane = planeReduction[0];
-		for( int i = 1; i < planeReduction.Size(); i++ ) {
-			plane += planeReduction[i];
-		}
-		plane.SetAt( problem.GetFeatureCount(), freeTerm );
-
-		CArray<double> distances;
-		distances.Add( 0.0, vectorCount );
-
-		NEOML_OMP_NUM_THREADS( curThreadCount )
-		{
-			int index = 0;
-			int count = 0;
-			if( OmpGetTaskIndexAndCount( vectorCount, index, count ) ) {
-				for( int i = 0; i < count; i++ ) {
-					CFloatVectorDesc desc;
-					matrix.GetRow( index, desc );
-					distances[index] = LinearFunction( plane, desc );
-					index++;
-				}
-			}
-		}
+		CCalcDistancesThreadTask distTask( params.ThreadCount, problem, plane );
```

Comment:
Better pass `threadPool->Size()` than `params.ThreadCount`.
It will be good if you change `ThreadCount` to `-1` or `0` in one of the tests of Svm